Code/Framework/AzFramework/AzFramework/Terrain/TerrainDataRequestBus.h

```diff
@@ -193,6 +195,116 @@ namespace AzFramework
             //! Given a ray, return the closest intersection with terrain.
             virtual RenderGeometry::RayResult GetClosestIntersection(const RenderGeometry::RayRequest& ray) const = 0;
 
+            //! A JobContext used to run jobs spawned by calls to the various Process*Async functions.
+            //! Instances of this class should only be created by calling CreateNewTerrainJobContext,
+            //! or by using the same JobManager as a previously created TerrainJobContext along with
+            //! a parent JobCancelGroup that is owned by another previously created TerrainJobContext,
+            //! and the only practical use for it is to cancel all Process*Async calls made using it.
+            class TerrainJobContext : public AZ::JobContext
+            {
+            public:
+                TerrainJobContext(AZ::JobManager& jobManager, AZ::JobCancelGroup& parentCancelGroup)
+                    : JobContext(jobManager)
+                    , m_terrainJobCancelGroup(&parentCancelGroup)
+                {
+                    SetCancelGroup(&m_terrainJobCancelGroup);
+                }
+
+                void Cancel() { m_terrainJobCancelGroup.Cancel(); }
+                bool IsCancelled() const { return m_terrainJobCancelGroup.IsCancelled(); }
+
+            private:
+                AZ::JobCancelGroup m_terrainJobCancelGroup;
+            };
+
+            //! Alias for an optional callback function to invoke when the various Process*Async functions complete.
+            //! The TerrainJobContext (if any) originally passed to the Process*Async function is passed as a param
+            //! to the callback function so it can be queried to see if the job was cancelled or completed normally.
+            typedef AZStd::function<void(AZStd::shared_ptr<TerrainJobContext>)> ProcessAsyncCompleteCallback;
+
+            //! A parameter group struct that can optionally be passed to the various Process*Async API functions.
+            struct ProcessAsyncParams
+            {
+                //! The TerrainJobContext which will run all jobs spawned by a call to a Process*Async function.
+                //! This must be set to a TerrainJobContext that was created by calling CreateNewTerrainJobContext,
+                //! but the same TerrainJobContext can be used for multiple calls to different Process*Async functions.
+                //! The only practical purpose of using this parameter is to be able to cancel all Process*Async calls
+                //! made using this TerrainJobContext object (which can be done by calling TerrainJobContext::Cancel()).
+                AZStd::shared_ptr<TerrainJobContext> m_terrainJobContext = nullptr;
```

Comment:
We could, but doing it like this means the calling code can make multiple async calls using the same TerrainJobContext and cancel them all at once.

Reply:
On the flip side, since the auto-created one _isn't_ returned right now, it's forcing the caller to allocate one just to be able to cancel at all.
I'm willing to wait and see how this plays out in the first API usage with the terrain debugger, but my gut tells me that shared job contexts will lead people astray - that if they're creating multiple requests, they probably will also want/need the ability to cancel individual requests from the set, not just do a bulk cancellation. I can see where bulk cancellation will _also_ be useful on things like component deactivates, but I suspect that will be more easily done through a loop of individual contexts rather than a single shared one.

Reply:
I'm fine changing it, it's cleaner than way.